File: ios/TacticalMaps/Map/GridMagnetic.swift
import Foundation

/// Grid-magnetic angle (grid north -> magnetic north) for the MGRS banner, so a
/// user can convert a grid bearing to a compass bearing. East positive.
///
/// G-M = magnetic declination - UTM grid convergence.
///  - declination (true -> magnetic) from the WMM2025 model in [GeoMagnetism] -
///    always available, no network, no magnetometer.
///  - convergence (true -> grid) = atan(tan(lon - centralMeridian) * sin(lat)),
///    the small rotation between true north and the UTM grid's north.
enum GridMagnetic {
    /// Preformatted like "G-M 13.4°E" / "G-M 2.1°W", or nil when there's no
    /// coordinate to compute for.
    static func label(latitude: Double?, longitude: Double?, altitudeMeters: Double = 0) -> String? {
        guard let lat = latitude, let lon = longitude else { return nil }
        let declination = GeoMagnetism.declinationDegrees(
            latitude: lat, longitude: lon, altitudeMeters: altitudeMeters)

        let zone = Int(floor((lon + 180.0) / 6.0)) + 1
        let centralMeridian = -180.0 + Double(zone - 1) * 6.0 + 3.0
        let convergence = atan(tan((lon - centralMeridian) * .pi / 180.0)
                               * sin(lat * .pi / 180.0)) * 180.0 / .pi

        let gm = declination - convergence
        return String(format: "G-M %.1f°%@", abs(gm), gm >= 0 ? "E" : "W")
    }
}

File: android/app/src/main/java/com/tacmap/map/GridMagnetic.kt
package com.tacmap.map

import android.hardware.GeomagneticField
import kotlin.math.abs
import kotlin.math.atan
import kotlin.math.floor
import kotlin.math.roundToInt
import kotlin.math.sin
import kotlin.math.tan

/**
 * Grid-magnetic angle (grid north -> magnetic north) for the MGRS banner, so a
 * user can convert a grid bearing to a compass bearing. East positive.
 *
 * G-M = magnetic declination - UTM grid convergence.
 *  - declination (true -> magnetic) comes from the platform's World Magnetic
 *    Model via [GeomagneticField] - always available, no network, no sensor.
 *  - convergence (true -> grid) = atan(tan(lon - centralMeridian) * sin(lat)),
 *    the small rotation between true north and the UTM grid's north.
 *
 * Returns the raw angle in degrees (+E / -W), or null when there's no
 * coordinate to compute for. [formatGridMagnetic] turns it into banner text.
 */
fun gridMagneticDegrees(lat: Double?, lon: Double?, altitudeMetres: Double? = null): Double? {
    if (lat == null || lon == null) return null
    val declination = GeomagneticField(
        lat.toFloat(), lon.toFloat(), (altitudeMetres ?: 0.0).toFloat(),
        System.currentTimeMillis()
    ).declination.toDouble()

    // UTM zone central meridian for this longitude.
    val zone = floor((lon + 180.0) / 6.0).toInt() + 1
    val centralMeridian = -180.0 + (zone - 1) * 6.0 + 3.0
    val convergence = Math.toDegrees(
        atan(tan(Math.toRadians(lon - centralMeridian)) * sin(Math.toRadians(lat)))
    )

    return declination - convergence
}

/**
 * Banner text for a G-M angle. Mils by default (NATO 6400/circle - what a
 * compass dial and a fire mission actually read); tapping the banner flips it
 * to degrees. E/W suffix. e.g. "G-M 222 mils E" or "G-M 12.5°E".
 */
fun formatGridMagnetic(degrees: Double, mils: Boolean): String {
    val dir = if (degrees >= 0) "E" else "W"
    return if (mils) {
        val m = (abs(degrees) * 6400.0 / 360.0).roundToInt()
        "G-M $m mils $dir"
    } else {
        "G-M %.1f°%s".format(abs(degrees), dir)
    }
}
